src/vm/perfmap.cpp

```diff
@@ -122,20 +155,195 @@ void PerfMap::Log(MethodDesc * pMethod, PCODE pCode, size_t codeSize)
         line.Printf("%p %x %s\n", pCode, codeSize, fullMethodSignature.GetANSI(scratch));
 
         // Write the line.
-        // The PAL already takes a lock when writing, so we don't need to do so here.
-        const char * strLine = line.GetANSI(scratch);
-        ULONG inCount = line.GetCount();
-        ULONG outCount;
-        m_FileStream->Write(strLine, inCount, &outCount);
+        WriteLine(line);
+    }
+    EX_CATCH{} EX_END_CATCH(SwallowAllExceptions);
+}
 
-        if (inCount != outCount)
+// Log a native image to the map.
+void PerfMap::LogNativeImage(PEFile * pFile)
+{
+    CONTRACTL{
+        THROWS;
+        GC_NOTRIGGER;
+        MODE_PREEMPTIVE;
+        PRECONDITION(pFile != NULL);
+    } CONTRACTL_END;
+
+    if (m_FileStream == NULL || m_ErrorEncountered)
+    {
+        // A failure occurred, do not log.
+        return;
+    }
+
+    // Logging failures should not cause any exceptions to flow upstream.
+    EX_TRY
+    {
+        // Get the native image name.
+        LPCUTF8 lpcSimpleName = pFile->GetSimpleName();
+
+        // Get the native image signature.
+        WCHAR wszSignature[39];
+        GetNativeImageSignature(pFile, wszSignature, lengthof(wszSignature));
+
+        SString strNativeImageSymbol;
+        strNativeImageSymbol.Printf("%s.ni.%S", lpcSimpleName, wszSignature);
+
+        // Get the base addess of the native image.
+        SIZE_T baseAddress =
+#ifndef DACCESS_COMPILE
```

Comment:
There should be one big #ifndef DACCESS_COMPILE around all logging methods instead.